pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,345 @@
+	clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := c.getCRD(clusterName, name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	logger := logging.WithObject(klog.FromContext(ctx), obj)
+	ctx = klog.NewContext(ctx, logger)
+
+	if !apihelpers.IsCRDConditionTrue(obj, apiextensionsv1.Established) {
+		// Exit early because binding controller waits on CRDs to be established before adding corresponding BoundResource
+		logger.V(4).Info("CRD is not established", "conditions", fmt.Sprintf("%#v", obj.Status.Conditions))
+		return nil
+	}
+
+	result, err := c.getAPIBindingsByBoundResourceUID(obj.Name)
+	if err != nil {
+		return err
+	}
+
+	if len(result) > 0 {
+		// An APIBinding that uses this bound CRD was found. Thus don't delete.
+		return nil
+	}
+
+	// Before this CRD can be deleted, check for any APIBindings that are in the process of being updated and requeue if any are found.
+	// This set of bindings could eventually refer to this CRD, but it hasn't been represented on the status subresource of the binding yet.
+	// This can happen in a couple of ways:
+	//
+	// 1. The first APIBinding for an APIExport was just created.
+	// 2. A new APIResourceSchema was added to an APIExport.
+	//
+	// In both scenarios, the CRD is created _before_ the APIBinding's status is updated.
+	//
+	// To catch "being updated" APIBindings in scenario 1, simply check if the phase is not "Bound".
+	// To catch "being updated" APIBindings in scenario 2, check the BindingUpToDate condition is False (since the phase will never change after initial bind).
+	//
+	// While the BindingUpToDate condition is actually set to False _after_ the CRD is created,
+	// the CRD established check earlier in this function should ensure that the condition has a chance to get updated before trying to (prematurely) delete this CRD.
+
+	bindingList, err := c.listAPIBindings()
+	if err != nil {
+		return err
+	}
+
+	for _, binding := range bindingList {
+		// Need to exclude APIBindings that have have invalid an APIExport reference.
+		// If these aren't ignored, then any single APIBinding with an invalid APIExport reference could prevent all bound CRDs from being deleted.
+		// Note that if an APIExport becomes invalid _after_ binding is complete, the BoundResources in the binding will still show up in the UID index (used above).
+		if conditions.IsFalse(binding, apisv1alpha1.APIExportValid) {
+			continue
+		}
+		if binding.Status.Phase != apisv1alpha1.APIBindingPhaseBound || !conditions.IsTrue(binding, apisv1alpha1.BindingUpToDate) {
+			uids, err := c.getResourceUIDs(binding)
+			if err != nil {
+				return err
+			}
+
+			for _, uid := range uids {
+				if string(uid) == obj.Name {
+					// Found an APIBinding that is still in the process of being bound that will eventually point to this CRD
+					return nil
+				}
+			}
+		}
+	}
```

Comment:
am still not a fan of this super complex lookup, even when optimized with an index. @ncdc any better idea?

Reply:
I tend to opt for a less correct, but good enough solution of just delaying deletion to give the controller time to converge.